kernel/src/process/signal/pauser.rs

```diff
@@ -112,159 +44,104 @@ impl Pauser {
-    fn new(
-        current_thread: Option<&'a Arc<Thread>>,
-        sig_mask: SigMask,
-        pauser: Arc<Pauser>,
-    ) -> Self {
-        let Some(thread) = current_thread.and_then(|thread| thread.as_posix_thread()) else {
-            return Self::KernelThread;
-        };
-
-        // Block `sig_mask`.
-        let (old_mask, filter) = {
-            let old_mask = thread.sig_mask().load(Ordering::Relaxed);
-            let new_mask = old_mask + sig_mask;
-            thread.sig_mask().store(new_mask, Ordering::Relaxed);
-
-            (old_mask, SigEventsFilter::new(new_mask))
-        };
-
-        // Register `SigQueueObserver`.
-        let observer = SigQueueObserver::new(pauser);
-        thread.register_sigqueue_observer(Arc::downgrade(&observer) as _, filter);
-
-        // Check pending signals after registering the observer to avoid race conditions.
-        if thread.has_pending() {
-            observer.set_interrupted();
-        }
-
-        Self::PosixThread {
-            thread,
-            old_mask,
-            observer,
-        }
-    }
-
-    fn is_interrupted(&self) -> bool {
-        match self {
-            Self::PosixThread { observer, .. } => observer.is_interrupted(),
-            Self::KernelThread => false,
+impl PauseInner for WaitQueue {
+    fn do_pause<F, R>(&self, mut cond: F, timeout: Option<&Duration>) -> Result<R>
+    where
+        F: FnMut() -> Option<R>,
+    {
+        if let Some(res) = cond() {
+            return Ok(res);
         }
-    }
-}
 
-impl<'a> Drop for SigObserverRegistrar<'a> {
-    fn drop(&mut self) {
-        let Self::PosixThread {
-            thread,
-            old_mask,
-            observer,
-        } = self
-        else {
-            return;
+        let (waiter, _) = Waiter::new_pair();
+        let cond = || {
+            self.enqueue(waiter.waker());
+            cond()
         };
 
-        // Restore the state, assuming no one else can modify the current thread's signal mask
-        // during the pause.
-        thread.unregiser_sigqueue_observer(&(Arc::downgrade(observer) as _));
-        thread.sig_mask().store(*old_mask, Ordering::Relaxed);
+        waiter.do_pause(cond, timeout)
     }
 }
 
-struct SigQueueObserver {
-    is_interrupted: AtomicBool,
-    pauser: Arc<Pauser>,
-}
-
-impl SigQueueObserver {
-    fn new(pauser: Arc<Pauser>) -> Arc<Self> {
-        Arc::new(Self {
-            is_interrupted: AtomicBool::new(false),
-            pauser,
-        })
-    }
-
-    fn is_interrupted(&self) -> bool {
-        self.is_interrupted.load(Ordering::Acquire)
-    }
+impl Pause for Waiter {}
+impl Pause for WaitQueue {}
 
-    fn set_interrupted(&self) {
-        self.is_interrupted.store(true, Ordering::Release);
-    }
-}
+/// Executes an operation while temporarily changing the signal mask of the current thread.
+pub fn with_sigmask<R>(mask: SigMask, operate: impl FnOnce() -> R) -> R {
+    let thread = current_thread!();
```

Comment:
The Pauser can also be used in Task Context, so use `Thread::current()` instead of `current_thread`

```rust
let Some(thread) = Thread::current() else {
    return None;
}
```